### ISSUE TITLE
Implement clustering accuracy

### ISSUE THREAD

**Issue by moetayuko** [Lightning-AI/torchmetrics#2767]
### Motivation

Clustering accuracy is a popular metric. In addition to classification accuracy, it employs the Hungarian algorithm to align the predicted pseudo labels and the ground truth labels.

Current implementations of clustering accuracy use either `scipy.optimize.linear_sum_assignment` or the [`munkres`](https://github.com/bmc/munkres) package for Hungarian. I'm not sure if this is allowed for torchmetrics, and a custom implementation needs to be added if not.

### Pitch

Implement clustering accuracy in `torchmetrics.clustering`

**Comment by SkafteNicki** [comment]
@moetayuko thanks for opening this issue. Do you have a reference to a source (possible research paper) where they describe the metric in details?

**Comment by moetayuko** [comment]
> @moetayuko thanks for opening this issue. Do you have a reference to a source (possible research paper) where they describe the metric in details?

Sec. 6.2.1 of https://arxiv.org/abs/2206.07579

FYI here are some random implementations I found:
https://github.com/bdy9527/SDCN/blob/da6bb007b7d07362ac04db3e146b4944e2acb883/evaluation.py#L9
https://github.com/google-research/google-research/blob/cc787a6a513cb6d2042cd6e286e2cbe42c41a863/stacked_capsule_autoencoders/capsules/eval.py#L29
https://github.com/aeon-toolkit/aeon/blob/01e8424df1bee9e614a25a25b5268383a8fd9893/aeon/performance_metrics/clustering.py#L12

**Comment by SkafteNicki** [comment]
@moetayuko thanks for the references, it really helped understanding how the metric is intended to work.
Hopefully, I have time to fully implement the metric in the next couple of days. I have already the logic figured out using https://github.com/ivan-chai/torch-linear-assignment for solving the linear sum assignment problem:
```python

from torchmetrics.functional.classification import multiclass_confusion_matrix
import torch
# pip install git+https://github.com/ivan-chai/torch-linear-assignment.git@main
from torch_linear_assignment import batch_linear_assignment

preds = torch.tensor([0, 0, 1, 1])
target = torch.tensor([1, 1, 0, 0])

confmat = multiclass_confusion_matrix(preds, target, num_classes=5)
print(confmat)

confmat = confmat[None]

assignment = batch_linear_assignment(confmat.max() - confmat)
print(assignment)

confmat = confmat[0]

tps = confmat[torch.arange(confmat.size(0)), assignment.flatten()]

acc = tps.sum() / len(preds)
print(acc)
```
